### PR TITLE
Avoid Two Sets of systemd Binaries in System Closure

### PR DESCRIPTION
**Pull request by blitz**
## Description of changes

We currently end up with two systemds in every system closure. One is the one you want and the second one is systemdMinimal. This includes the full set of systemd binaries, but it's only there to provide libudev and libsystemd. This wastes around 10 MiB of space.

I'm not sure that what I've done here is a good idea. Consider it a PoC.

Also see #261795.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/systemd/default.nix [hunk 1]
```diff
@@ -697,7 +699,13 @@ stdenv.mkDerivation (finalAttrs: {
     export DESTDIR=/
   '';
 
-  postInstall = ''
+  installPhase = ''
+    runHook preInstall
+    meson install ${lib.optionalString onlyLibs "--tags devel,libudev,libsystemd"}
```
Comment (blitz) [inline pkgs/os-specific/linux/systemd/default.nix:704]: This should probably go into `pkgs/development/tools/build-managers/meson/setup-hook.sh`.

**Comment by arianvp** [comment]
I like the idea of this change. I think we can even delete `systemdMinimal` now. It was only there for breaking the dependency loop between `systemd -> lvm2 -> udev`. If  the `systemdLibs` package has no circular dependency on lvm2 then this is a much more elegant solution

File: pkgs/os-specific/linux/systemd/default.nix [hunk 2]
```diff
@@ -435,7 +437,7 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optional withUkify (python3Packages.python.withPackages (ps: with ps; [ pefile ]))
   ;
 
-  #dontAddPrefix = true;
+  mesonBuildType = "release";
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/systemd/default.nix:440]: Isn't this the default?

Comment (blitz) [reply pkgs/os-specific/linux/systemd/default.nix:440]: The default is 'plain'. I wonder whether we should change it.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/systemd/default.nix [hunk 3]
```diff
@@ -697,7 +699,13 @@ stdenv.mkDerivation (finalAttrs: {
     export DESTDIR=/
```
[the remark this answers is not in the document]
Comment (arianvp) [reply pkgs/os-specific/linux/systemd/default.nix:699]: It is Needed. Please keep reviews on topic and only on the patch at hand

File: pkgs/os-specific/linux/systemd/default.nix [hunk 4]
```diff
@@ -371,7 +373,7 @@ stdenv.mkDerivation (finalAttrs: {
     patchShebangs tools test src/!(rpm|kernel-install|ukify) src/kernel-install/test-kernel-install.sh
   '';
 
-  outputs = [ "out" "man" "dev" ];
+  outputs = [ "out" "dev" ] ++ (lib.optional (!onlyLibs) "man");
```
[the remark this answers is not in the document]
Comment (blitz) [reply pkgs/os-specific/linux/systemd/default.nix:376]: The man pages output ends up being empty and the build logic doesn't like that. The systemd build system also has no tag to install the man pages, as far as I can see.

File: doc/hooks/meson.section.md [hunk 5]
```diff
@@ -12,11 +12,11 @@ By default, `enableParallelBuilding` is enabled as Meson supports parallel build
 
 Controls the flags passed to `meson setup`.
 
-##### `mesonCheckFlags` {#mesoncheckflags}
```
Comment (blitz) [inline doc/hooks/meson.section.md:15]: The headers where `###` and `####` for no apparent reason. I fixed them to be consistent.

**Comment by blitz** [comment]
@arianvp 

> I think we can even delete `systemdMinimal` now.

The problem is that some places actually use this as a "minimal systemd" and not as "libudev". :(

File: pkgs/os-specific/linux/systemd/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/os-specific/linux/systemd/default.nix]: OK.

**Comment by AndersonTorres** [comment]
> The problem is that some places actually use this as a "minimal systemd" and not as "libudev". :(

Like an alias?

File: doc/hooks/meson.section.md (hunk 5 above)
Comment (AndersonTorres) [reply doc/hooks/meson.section.md:15]: Do this in another pull request. It's better not to mix things.

**Comment by AndersonTorres** [comment]
You wrote on the commit message:

> Projects building with meson are currently installTargets. Map these to install tags, which are roughly equivalent. This allows projects to selectively install components.

A not so long time ago I came across a nasty bug: I treated `wafConfigureFlags` as _roughly equivalent_ to `configureFlags`. It caused a problem because some configureFlags are not recognized by waf.

https://github.com/NixOS/nixpkgs/pull/254111

Since then I do not recommend the reuse of flags.

So, I suggest an exclusive name for this:  `mesonInstallTags`.

Also, in tandem I will open a refactor PR of Meson, specifically capturing this issue.

**Comment by flokli** [comment]
> The problem is that some places actually use this as a "minimal systemd" and not as "libudev". :(

Then this should be fixed :-)

**Comment by blitz** [comment]
@AndersonTorres 

> So, I suggest an exclusive name for this: `mesonInstallTags`.
> 
> Also, in tandem I will open a refactor PR of Meson, specifically capturing this issue.

If you do this, that'd be great. I can also change it here for now.

**Comment by blitz** [comment]
@flokli 

> > The problem is that some places actually use this as a "minimal systemd" and not as "libudev". :(
> 
> Then this should be fixed :-)

This PR sort of does this. `systemdMinimal` is now the minimal systemd people hope for and for dependency breaking there is `systemdLibs`.

File: pkgs/development/tools/build-managers/meson/setup-hook.sh [hunk 6]
```diff
@@ -38,6 +38,11 @@ mesonConfigurePhase() {
     runHook postConfigure
 }
 
+if [ -z "${dontUseMesonConfigure-}" -a -z "${configurePhase-}" ]; then
```
Comment (AndersonTorres) [inline pkgs/development/tools/build-managers/meson/setup-hook.sh:41]: Avoid stylistical reformatting

Comment (blitz) [reply pkgs/development/tools/build-managers/meson/setup-hook.sh]: Fixed.

**Comment by blitz** [comment]
@AndersonTorres I've resolved the issues. Do you think it makes sense to fix existing uses of `systemdMinimal` in this PR (use `systemdLibs` where appropriate) or would this be better in a subsequent PR?

**Comment by AndersonTorres** [comment]
Another PR is better.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -28837,12 +28827,15 @@ with pkgs;
...
+  };
+  systemdLibs = systemdMinimal.override {
+    pname = "systemd-minimal-libs";
+    buildLibsOnly = true;
   };
 
   udev =
     if (with stdenv.hostPlatform; isLinux && isStatic) then libudev-zero
-    else systemd; # TODO: change to systemdMinimal
+    else systemdLibs;
```
[the remark this answers is not in the document]
Comment (arianvp) [reply pkgs/top-level/all-packages.nix:28838]: @blitz thoughts?

**Comment by arianvp** [comment]
I think we should revert this PR. A lot of places in nixpkgs assume udevadm to exist in the udev package

If you search for udevadm there are a lot of hits

File: pkgs/top-level/all-packages.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (blitz) [reply pkgs/top-level/all-packages.nix:28838]: @arianvp Let's revert this change then. Long-term I think the better solution is to create a real udev package with libs and binaries and all. The systemd build system makes this hard though. I have to study how Fedora does this.

[the remark this answers is not in the document]
Comment (arianvp) [reply pkgs/top-level/all-packages.nix:28838]: This was done as an alternative: https://github.com/NixOS/nixpkgs/pull/268718

[the remark this answers is not in the document]
Comment (blitz) [reply pkgs/top-level/all-packages.nix:28838]: Ok. That also makes sense.

If it still comes to reverting this, please leave the `mesonInstallTags` change in place. I think this is genuinely useful elsewhere.

**Pull request merged**